rest/oidc_test_provider_test.go

```diff
@@ -39,3 +48,85 @@ func TestExtractSubjectFromRefreshToken(t *testing.T) {
+		},
+	}
+	defaultProvider := "test"
+	opts := auth.OIDCOptions{Providers: providers, DefaultProvider: &defaultProvider}
+	restTesterConfig := RestTesterConfig{
+		DatabaseConfig: &DbConfig{
+			OIDCConfig: &opts,
+			Unsupported: db.UnsupportedOptions{
+				OidcTestProvider: db.OidcTestProviderOptions{
+					Enabled: true,
+				},
+				OidcTlsSkipVerify: true,
+			},
+		}}
+	restTester := NewRestTester(t, &restTesterConfig)
+	restTester.SetAdminParty(false)
+	defer restTester.Close()
+
+	mockSyncGateway := httptest.NewTLSServer(restTester.TestPublicHandler())
+	defer mockSyncGateway.Close()
+	mockSyncGatewayURL := mockSyncGateway.URL
+	provider := restTesterConfig.DatabaseConfig.OIDCConfig.Providers.GetDefaultProvider()
+	provider.Issuer = mockSyncGateway.URL + "/db/_oidc_testing"
+	provider.CallbackURL = base.StringPtr(mockSyncGateway.URL + "/db/_oidc_callback")
+
+	// Send OpenID Connect request
+	authURL := "/db/_oidc?provider=test&offline=true"
+	requestURL := mockSyncGatewayURL + authURL
+	request, err := http.NewRequest(http.MethodGet, requestURL, nil)
+	require.NoError(t, err, "Error creating new request")
+	jar, err := cookiejar.New(nil)
+	require.NoError(t, err, "Error creating new cookie jar")
+	client := auth.GetHttpClient(true)
+	client.Jar = jar
+	response, err := client.Do(request)
+	require.NoError(t, err, "Error sending request")
+	require.Equal(t, http.StatusOK, response.StatusCode)
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	require.NoError(t, err, "Error reading response")
+	bodyString := string(bodyBytes)
+	require.NoError(t, response.Body.Close(), "Error closing response body")
+
+	// Send authentication request
+	requestURL = mockSyncGateway.URL + "/db/_oidc_testing/" + parseAuthURL(bodyString)
+	form := url.Values{}
+	form.Add("username", "alice")
+	form.Add("authenticated", "Return a valid authorization code for this user")
+	request, err = http.NewRequest(http.MethodPost, requestURL, bytes.NewBufferString(form.Encode()))
+	require.NoError(t, err, "Error creating new request")
+	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	response, err = client.Do(request)
+	require.NoError(t, err, "Error sending request")
+	require.Equal(t, http.StatusOK, response.StatusCode)
+
+	var authResponseActual OIDCTokenResponse
+	require.NoError(t, err, json.NewDecoder(response.Body).Decode(&authResponseActual))
+	require.NoError(t, response.Body.Close(), "Error closing response body")
+	assert.NotEmpty(t, authResponseActual.SessionID, "session_id doesn't exist")
+	assert.NotEmpty(t, authResponseActual.Username, "session_id doesn't exist")
+	assert.NotEmpty(t, authResponseActual.IDToken, "id_token mismatch")
+	assert.NotEmpty(t, authResponseActual.RefreshToken, "refresh_token mismatch")
+}
+
+// parseAuthURL returns the authentication URL extracted from user consent form.
+func parseAuthURL(html string) string {
+	re := regexp.MustCompile(`<form action=".*?(.*)" method`)
```

Comment:
Minor suggestion on this to simplify.
We want to ensure we match on the whole value of action, not just part of it. (i.e. we don't want the `.*?` accidentally matching `authenticate`, leaving it out of the group)

```suggestion
	re := regexp.MustCompile(`<form action="(.+)" method`)
```

Reply:
Sure, this is a good idea.